### ISSUE TITLE
"sha256sum or shasum is not available for use" error installing nodejs

### ISSUE THREAD

**Issue by markalfred** [asdf-vm/asdf-nodejs#137]
# TL;DR
When running `asdf install nodejs`, if you see **"sha256sum or shasum is not available for use"**, run `which shasum`.

If asdf reports **"asdf: No version set for command which"**, you should `npm install -g which` then `asdf install nodejs` again.

# The Issue

Trying to install nodejs from my home directory:
```
~ ⟩ asdf install nodejs 12.6.0

gpg: Signature made Wed Jul  3 09:55:58 2019 EDT
gpg:                using RSA key 8FCCA13FEF1D0C2E91008E09770F7A9A5AE15600
gpg: Good signature from "Michaël Zasso (Targos) <targos@protonmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 8FCC A13F EF1D 0C2E 9100  8E09 770F 7A9A 5AE1 5600
sha256sum or shasum is not available for use
Authenticity of package archive can not be assured. Exiting.
```

As far as I know, `shasum` should exist, so I check where it's located:

```
~ ⟩ which -a shasum

asdf: No version set for command which
you might want to add one of the following in your .tool-versions file:
nodejs 8.2.1
```

It turns out that asdf has blocked me from using my system's `which` command, because [`node-which`](https://github.com/npm/node-which) is a dependency of some package in one of my projects.

```
~ ⟩ cd MyProject

~/MyProject ⟩ which -a which

~/.asdf/installs/nodejs/8.2.1/.npm/bin/which
~/MyProject/node_modules/.bin/which
~/.asdf/shims/which
/usr/bin/which
```

Cool, so inside of this project's directory, maybe asdf will be able to find `shasum`, since it can use `node-which` to find it:

```
~/MyProject ⟩ asdf install nodejs 12.6.0

gpg: Signature made Wed Jul  3 09:55:58 2019 EDT
gpg:                using RSA key 8FCCA13FEF1D0C2E91008E09770F7A9A5AE15600
gpg: Good signature from "Michaël Zasso (Targos) <targos@protonmail.com>" [ultimate]
node-v12.6.0-darwin-x64.tar.gz: OK
```

😭 

It seems presumptuous of asdf to assume that because it knows of a package with a bin named `which`, I should be forced to install that package in whichever node version I'm currently using. Some packages may be named the same as system packages, like this one, and if asdf can't find the node package, maybe it should fall back to the system one instead.

I'm not sure, but I've seen other people posting issues installing node, and sometimes they have the same error message: "sha256sum or shasum is not available for use", but I haven't seen this solution mentioned anywhere.

**Comment by Stratus3D** [comment]
`type -a which` is very helpful in situations like this.

**Comment by markalfred** [comment]
@Stratus3D agreed, however the only indication that something wasn't working correctly was this

> sha256sum or shasum is not available for use

and it's not obvious from that error message that the issue was with `which` — it seems more like an issue of `shasum`.

**Comment by Stratus3D** [comment]
@markalfred I'm not sure what the best approach here is. I suppose we should fall back to the system installed `which` when no version is selected for the asdf-installed `which`. This isn't a feature asdf supports yet, so in the meantime I would recommend uninstall the which npm package and running `asdf reshim`. The shim should get deleted and allow you to access the system which.

**Comment by markalfred** [comment]
@Stratus3D `which` isn't actually a dependency of any of my projects, it's deep in the dependency tree:

```
> npm ls which
├─┬ @storybook/cli@3.2.12
│ ├─┬ child-process-promise@2.2.1
│ │ └─┬ cross-spawn@4.0.2
│ │   └── which@1.3.0  deduped
│ └─┬ cross-spawn@5.1.0
│   └── which@1.3.0
├─┬ babel-jest@24.8.0
│ └─┬ @jest/transform@24.8.0
│   └─┬ jest-haste-map@24.8.0
│     └─┬ sane@4.1.0
│       └─┬ execa@1.0.0
│         └─┬ cross-spawn@6.0.5
│           └── which@1.3.0  deduped
├─┬ jest@24.8.0
│ └─┬ jest-cli@24.8.0
│   ├─┬ @jest/core@24.8.0
│   │ ├─┬ @jest/reporters@24.8.0
│   │ │ └─┬ node-notifier@5.4.0
│   │ │   └── which@1.3.1
│   │ └─┬ jest-changed-files@24.8.0
│   │   └─┬ execa@1.0.0
│   │     └─┬ cross-spawn@6.0.5
│   │       └── which@1.3.0  deduped
│   └─┬ yargs@12.0.5
│     └─┬ os-locale@3.1.0
│       └─┬ execa@1.0.0
│         └─┬ cross-spawn@6.0.5
│           └── which@1.3.0  deduped
└─┬ loki@0.6.5
  ├─┬ gm@1.23.1
  │ └─┬ cross-spawn@4.0.2
  │   └── which@1.3.0  deduped
  └─┬ osnap@1.1.0
    └── which@1.3.0  deduped
```

I think the two workarounds for me, and anyone else who runs into this, are:

1. Install `which` globally, then update asdf
2. cd to a project directory that contains `which`, then update asdf

Neither of these is too much of a hassle, it was just a hassle to figure out what the underlying problem was. One way that asdf might make it easier to see the problem would be to bubble up the error message if `which shasum` results in an error. Something like...

```
sha256sum or shasum is not available for use - received the following error:
    No version set for command which
    you might want to add one of the following in your .tool-versions file:
    nodejs 8.2.1
Authenticity of package archive can not be assured. Exiting.
```

**Comment by Stratus3D** [comment]
So as a workaround, since you are just trying to run `asdf install nodejs 12.6.0`, you could set your node version to `system`, to allow use of all system executables, which should give the `asdf install nodejs 12.6.0` command access to the real `which`, which should allow the install command to succeed.

There isn't a proper solution to this problem right now, and there probably won't be something released for a while.

**Comment by lnikkila** [comment]
I ran into this same issue recently.

`command -v` is a POSIX-compliant way to check whether something exists in PATH and might be better in this case, it's a shell builtin in bash:

```sh
if command -v shasum > /dev/null; then
  # shasum exists
fi
```
